k9mail/src/main/java/security/planck/messaging/MessagingRepository.kt

```diff
@@ -0,0 +1,269 @@
+    private val appIoScope: CoroutineScope,
+    private val dispatcherProvider: DispatcherProvider,
+) {
+
+    suspend fun loadMessage(
+        messageViewUpdate: MessageViewUpdate,
+    ) = withContext(dispatcherProvider.io()) {
+        messageViewUpdate.stateFlow.value = MessageViewState.Loading
+        messageViewUpdate.effectFlow.value = MessageViewEffect.NoEffect
+        loadMessageFromDatabase(
+            messageViewUpdate
+        ).join()
+    }
+
+    private fun loadMessageFromDatabase(
+        messageViewUpdate: MessageViewUpdate,
+    ) = appIoScope.launch {
+        with(messageViewUpdate) {
+            kotlin.runCatching {
+                controller.loadMessage(account, messageReference.folderName, messageReference.uid)
+            }.onFailure {
+                Timber.e(it)
+                effectFlow.value = MessageViewEffect.ErrorLoadingMessage(it)
+            }.onSuccess { message ->
+                message?.let {
+                    messageLoaded(message, messageViewUpdate)
+                } ?: let { effectFlow.value = MessageViewEffect.ErrorLoadingMessage() }
+            }
+        }
+    }
+
+    private suspend fun messageLoaded(
+        message: LocalMessage,
+        messageViewUpdate: MessageViewUpdate,
+    ) {
+        message.recoverRating()
+        val loadedState = if (!message.hasToBeDecrypted()) {
+            MessageViewState.DecryptedMessageLoaded(message)
+        } else {
+            MessageViewState.EncryptedMessageLoaded(message)
+        }
+
+        with(messageViewUpdate) {
+            withContext(dispatcherProvider.main()) {
+                stateFlow.value = loadedState
+            }
+            if (message.isMessageIncomplete()) {
+                downloadMessageBody(messageViewUpdate, false)
+            } else if (message.hasToBeDecrypted()) {
+                decryptMessage(message, messageViewUpdate)
+            } else {
+                decodeMessage(message, messageViewUpdate)
+            }
+        }
+    }
+
+    private fun decodeMessage(
+        message: LocalMessage,
+        messageViewUpdate: MessageViewUpdate,
+    ) {
+        with(messageViewUpdate) {
+            kotlin.runCatching {
+                infoExtractor.extractMessageForView(
+                    message,
+                    null,
+                    account.isOpenPgpProviderConfigured
+                )
+            }.onFailure {
+                Timber.e(it)
+                stateFlow.value =
+                    MessageViewState.ErrorDecodingMessage(
+                        createErrorStateMessageViewInfo(message),
+                        it
+                    )
+            }.onSuccess {
+                stateFlow.value = MessageViewState.MessageDecoded(it)
+            }
+        }
+    }
+
+    private fun downloadMessageBody(
+        messageViewUpdate: MessageViewUpdate,
+        complete: Boolean,
+    ) {
+        with(messageViewUpdate) {
+            if (complete) {
```

Comment:
could be replaced with WHEN
that can add flexibility for supporting more conditions later on